### DISCUSSION TITLE
Zustand persist with Expo GO

### DISCUSSION THREAD

**Discussion by Revaycolizer** [pmndrs/zustand#2411] [General]
I am using Zustand to handle state management in my expo project, however in development using expo go app, the state is handled successfully, when I exit expo go and open it again it redirects to the home page as the state is successfully present, but after building apk, opening the app, sign in ,exit the app and open it again a user has to sign in again, how can I fix this  

```ts
import AsyncStorage from "@react-native-async-storage/async-storage";
import {create} from "zustand";

import {
  createJSONStorage,
  persist,
} from 'zustand/middleware'

type userAction={
  setUser:(user:User['user'])=>void
}

export const useUser = create<User & userAction>()(
  persist(
    (set) => ({
      user:null,
      setUser: (user: string) => set(() => ({ user: user})),
    }),
    {
      name: 'user-storage',
      storage: createJSONStorage(() =>  AsyncStorage),
    }
  )
)```

**Comment by dhruvwill** [comment]
I think by default, even if zustand is combined with persist middleware, we still have to use some disk storage mechanism such as AsyncStorage or MMKV. 
You should combine zustand with MMKV.

**Reply by Revaycolizer** [reply]
I used Asyncstorage but still nothing

**Reply by Revaycolizer** [reply]
```export const useUser = create<User & userAction>()(
  persist(
    (set) => ({
      user:null,
      setUser: (user: string) => set(() => ({ user: user})),
    }),
    {
      name: 'user-storage',
      storage: createJSONStorage(() =>  AsyncStorage),
    }
  )
)```

**Reply by dhruvwill** [reply]
can you please share your github repository, then maybe I can have a look. I did this today only, the same way, it worked.
Extras:- use Async Api to make a component which retrieves all keys from AsyncStorage, this will help you see the contents of AsyncStorage as a whole.

**Reply by dbritto-dev** [reply]
@Revaycolizer I encourage you to create a custom storage and pass it to `createJSONStorage(() => storage)` instead of `createJSONStorage(() => AsyncStorage)`

**Reply by quememo** [reply]
> @Revaycolizer I encourage you to create a custom storage and pass it to `createJSONStorage(() => storage)` instead of `createJSONStorage(() => AsyncStorage)`

why?

**Reply by dbritto-dev** [reply]
> > @Revaycolizer I encourage you to create a custom storage and pass it to `createJSONStorage(() => storage)` instead of `createJSONStorage(() => AsyncStorage)`
> 
> 
> 
> why?

You need to think on few things:
- the storage use localStorage which is actually an instance of Storage
- Using static method from a static class is not same as using methods from an object
- sometimes async storages needs to load before being available, so you need to check if is available or not
- you can replace localStorage with something else but is not like a straightforward implementation so you need to add some checks

In a nutshell is better to pass a storage instance (object instance) than a static class

**Comment by pedroafsouza** [comment]
I created a package called expo-zustland-persist. This one is compatible with react-native and expo.


Here is the link => https://www.npmjs.com/package/expo-zustand-persist

**Reply by Revaycolizer** [reply]
> I created a package called expo-zustland-persist. This one is compatible with react-native and expo.
> 
> 
> Here is the link => https://www.npmjs.com/package/expo-zustand-persist

Thanks seems promising I'll give it a try
